### PR TITLE
[C-Api] prepare new APIs

### PR DESCRIPTION
**Pull request by jaeyun-jung**
prepare next ACR
- function to invoke with preallocated output handle
- function to open nn model with custom options

Change function names and add description for each API.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

File: api/capi/include/nnstreamer-single.h [hunk 1]
```diff
@@ -106,6 +141,23 @@ int ml_single_close (ml_single_h single);
  */
 int ml_single_invoke (ml_single_h single, const ml_tensors_data_h input, ml_tensors_data_h *output);
 
+/**
+ * @brief Invokes the model with the given input data and fills the @a output data handle.
+ * @details This function does not allocate new data handle, unlike ml_single_invoke() which returns newly allocated output handle.
+ * @since_tizen 6.5
+ * @param[in] single The model handle to be inferred.
+ * @param[in] input The input data to be inferred.
+ * @param[in] output The output data to be filled.
```
Comment (myungjoo) [inline api/capi/include/nnstreamer-single.h:150]: Need more info for developers:
```
The output data to be filled.
```
==>
```
The output data to be filled by the API. Output should be preallocated before calling the API.
```

File: api/capi/include/nnstreamer-single.h [hunk 2]
```diff
@@ -106,6 +141,23 @@ int ml_single_close (ml_single_h single);
  */
 int ml_single_invoke (ml_single_h single, const ml_tensors_data_h input, ml_tensors_data_h *output);
 
+/**
+ * @brief Invokes the model with the given input data and fills the @a output data handle.
```
Comment (myungjoo) [inline api/capi/include/nnstreamer-single.h:145]: Add more description
```
Invokes the model with the given input data and fills the @a output data handle.
```
==>
```
Invokes the model with the given input data and fills the @a output data handle. This is generally faster than ml_invoke_single.
```

File: api/capi/include/nnstreamer-single.h [hunk 3]
```diff
@@ -106,6 +141,23 @@ int ml_single_close (ml_single_h single);
...
+ * @details This function does not allocate new data handle, unlike ml_single_invoke() which returns newly allocated output handle.
```
Comment (myungjoo) [inline api/capi/include/nnstreamer-single.h:146]:
```
This function does not allocate new data handle, unlike ml_single_invoke() which returns newly allocated output handle.
```
==>
```
The caller should preallocate memory buffers of the given output handle before calling the API.
Note that ml_single_invoke() allocates memory buffers of the output handle in the API, which may incur memcpy.
```

**Review by myungjoo — COMMENTED** [review]
LGTM. Please rephrase a few doxygen entries.

File: api/capi/include/nnstreamer-single.h (hunk 1 above)
Comment (again4you) [inline api/capi/include/nnstreamer-single.h:150]: Maybe [in/out]?
`@param[in] output` => `@param[in/out] output`

**Review by again4you — APPROVED** [review]
Except comments, LGTM!

File: api/capi/include/nnstreamer-single.h (hunk 1 above)
Comment (kparichay) [reply api/capi/include/nnstreamer-single.h]: Addon : `The memory of each output must be contiguous. However, it is not required for multiple outputs to have contiguous shared memory.`?

**Comment by kparichay** [comment]
This is most likely outside the scope of this PR but I want your opinion.
This patch will allow supporting pre-allocated handle to pass to `invoke`.
However, as nnstreamer `tensor_data` c-api does not allow to create a wrapper of `tensor_data` around already allocated memory, I think the usage of this API might be limited. 
How about introducing an additional API like `int ml_tensors_data_create (const ml_tensors_info_h info, ml_tensors_data_h *handle, void *data)` where `void *data` is the pre-allocated memory allocated by the user. This function would just wrap the user allocated memory with `tensor_data` so that it can be used inside nnstreamer. (I know this is currently limited to just 1 tensor but can be extended)

**Comment by jaeyun-jung** [comment]
@kparichay I added new issue about your suggestion. (#2827)

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**